crossvalidation.py

```diff
@@ -132,12 +149,31 @@ def objective(params, boundary, data, W, results_file=None, tuning_params=None):
         print(f"Done: test error {test_error}")
 
 
-    if results_file is not None:
+    # get mean test error
+    mean_errs = np.mean(errs)
+
+    if track_error is not None:
+        # update if new test was better
+        track_error.update(mean_errs)
+
+    if track_error is not None and track_error.updated:
         print(f'saving file {results_file}')
         with open(results_file, 'wb') as file:
             pickle.dump(results, file, protocol=pickle.HIGHEST_PROTOCOL)
-    
-    return np.mean(errs)
+        
+        print(f'saving new best params to {best_params_file}')
+        tuning_dict.update({"mean_error": track_error.test_error})
+        with open(best_params_file, 'w') as file:
+            json.dump(tuning_dict, file)
+        track_error.updated = False
+
+    if tuning_params is None:
+        print(f'saving file {results_file}')
+        with open(results_file, 'wb') as file:
+            pickle.dump(results, file, protocol=pickle.HIGHEST_PROTOCOL)
+
+        
+    return mean_errs
```

Comment:
This section is all a bit clunky, I'm happy to refactoring if you think it needs it.